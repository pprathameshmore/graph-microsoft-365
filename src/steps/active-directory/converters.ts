import {
  createIntegrationEntity,
  createDirectRelationship,
  Entity,
  getTime,
  IntegrationInstance,
  Relationship,
  RelationshipDirection,
  RelationshipClass,
  createMappedRelationship,
} from '@jupiterone/integration-sdk-core';
import { Group, Organization, User } from '@microsoft/microsoft-graph-types';

import { GroupMember, MemberType } from './clients/directoryClient';
import { entities, relationships } from './constants';

export function createAccountEntity(instance: IntegrationInstance): Entity {
  return createIntegrationEntity({
    entityData: {
      source: {},
      assign: {
        _class: entities.ACCOUNT._class,
        _key: `${entities.ACCOUNT._type}-${instance.id}`,
        _type: entities.ACCOUNT._type,
        name: instance.name,
        displayName: instance.name,
      },
    },
  });
}

export function createAccountEntityWithOrganization(
  instance: IntegrationInstance,
  organization: Organization,
  intuneConfig: {
    mobileDeviceManagementAuthority?: string;
    subscriptionState?: string;
    intuneAccountID?: string;
  },
): Entity {
  let defaultDomain: string | undefined;
  const verifiedDomains = organization.verifiedDomains?.map((e) => {
    if (e.isDefault) {
      defaultDomain = e.name as string | undefined;
    }
    return e.name as string;
  });

  return createIntegrationEntity({
    entityData: {
      source: {
        organization,
        intuneConfig,
      },
      assign: {
        _class: entities.ACCOUNT._class,
        _key: `${entities.ACCOUNT._type}-${instance.id}`,
        _type: entities.ACCOUNT._type,
        id: organization.id,
        name: organization.displayName,
        displayName: instance.name,
        organizationName: organization.displayName,
        defaultDomain,
        verifiedDomains,
        intuneAccountId: intuneConfig?.intuneAccountID,
        mobileDeviceManagementAuthority:
          intuneConfig?.mobileDeviceManagementAuthority,
        intuneSubscriptionState: intuneConfig?.subscriptionState,
      },
    },
  });
}

export function createGroupEntity(data: Group): Entity {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: entities.GROUP._class,
        _type: entities.GROUP._type,
        name: data.displayName,
        displayName: data.displayName as string | undefined,
        id: data.id,
        deletedOn: getTime(data.deletedDateTime),
        classification: data.classification,
        createdOn: getTime(data.createdDateTime),
        description: data.description,
        email: data.mailEnabled ? data.mail : undefined,
        mail: data.mailEnabled ? data.mail : undefined,
        mailEnabled: data.mailEnabled,
        mailNickname: data.mailNickname,
        renewedOn: getTime(data.renewedDateTime),
        securityEnabled: data.securityEnabled,
      },
    },
  });
}

export function generateUserKey(user: User): string {
  return user.id as string;
}

export function createUserEntity(data: User): Entity {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _key: generateUserKey(data),
        _class: entities.USER._class,
        _type: entities.USER._type,
        name: data.displayName,
        username: data.userPrincipalName,
        displayName: data.displayName as string | undefined,
        givenName: data.givenName,
        firstName: data.givenName,
        jobTitle: data.jobTitle,
        email: data.mail ?? undefined,
        mail: data.mail ?? undefined,
        mobilePhone: data.mobilePhone,
        officeLocation: data.officeLocation,
        preferredLanguage: data.preferredLanguage,
        surname: data.surname,
        lastName: data.surname,
        userPrincipalName: data.userPrincipalName,
        id: data.id,
      },
    },
  });
}

export function createOrganizationEntity(data: Organization): Entity {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: entities.ORGANIZATION._class,
        _type: entities.ORGANIZATION._type,
        website: data.verifiedDomains?.map((domain) => domain.name).join(', '),
        emailDomain: data.verifiedDomains
          ?.filter((domain) => domain.capabilities?.includes('Email'))
          .map((domain) => domain.name)
          .join(', '),
        external: undefined,
      },
    },
  });
}

export function createAccountGroupRelationship(
  account: Entity,
  group: Entity,
): Relationship {
  return createDirectRelationship({
    _class: RelationshipClass.HAS,
    fromKey: account._key,
    fromType: entities.ACCOUNT._type,
    toKey: group.id as string,
    toType: entities.GROUP._type,
    properties: {
      _type: relationships.ACCOUNT_HAS_GROUP._type,
    },
  });
}

export function createAccountUserRelationship(
  account: Entity,
  user: Entity,
): Relationship {
  return createDirectRelationship({
    _class: RelationshipClass.HAS,
    from: account,
    to: user,
  });
}

export function createGroupMemberRelationship(
  group: Entity,
  member: GroupMember,
): Relationship {
  const memberEntityType = getGroupMemberEntityType(member);
  const memberEntityClass = getGroupMemberEntityClass(member);

  return createMappedRelationship({
    _class: RelationshipClass.HAS,
    _type: relationships.GROUP_HAS_MEMBER._type,
    _mapping: {
      relationshipDirection: RelationshipDirection.FORWARD,
      sourceEntityKey: group._key,
      targetFilterKeys: [['_type', '_key']],
      targetEntity: {
        _key: member.id,
        _type: memberEntityType,
        _class: memberEntityClass,
        displayName: member.displayName,
        jobTitle: member.jobTitle,
        email: member.mail,
      },
    },
    properties: {
      groupId: group.id as string,
      memberId: member.id,
      memberType: member['@odata.type'],
    },
  });
}

function getGroupMemberEntityType(member: GroupMember): string {
  switch (member['@odata.type']) {
    case MemberType.USER:
      return entities.USER._type;
    case MemberType.GROUP:
      return entities.GROUP._type;
    default:
      return entities.GROUP_MEMEBER._type;
  }
}

function getGroupMemberEntityClass(member: GroupMember): string | string[] {
  switch (member['@odata.type']) {
    case MemberType.USER:
      return entities.USER._class;
    case MemberType.GROUP:
      return entities.GROUP._class;
    default:
      return entities.GROUP_MEMEBER._class;
  }
}
