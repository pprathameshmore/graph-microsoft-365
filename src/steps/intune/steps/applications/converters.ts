import {
  createIntegrationEntity,
  Entity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
import {
  AndroidManagedStoreApp,
  ManagedApp,
  WebApp,
  IosLobApp,
  AndroidLobApp,
  WindowsPhoneXAP,
  MobileLobApp,
  DetectedApp,
} from '@microsoft/microsoft-graph-types-beta';
import { entities } from '../../constants';

export const MANAGED_APP_KEY_PREFIX = 'IntuneManaged:';
export const DETECTED_APP_KEY_PREFIX = 'IntuneDetected:';
export const UNVERSIONED = 'unversioned';

/**
 * Creates an application entity that represents an application that is being managed by Intune.
 * This should not contain information unique to specific device installations.
 * Intune uses this to apply policies and configurations to applications either off the shelf or uploaded to Intune.
 * https://docs.microsoft.com/en-us/graph/api/resources/intune-apps-managedapp?view=graph-rest-beta
 */
export function createManagedApplicationEntity(
  managedApp: ManagedApp & { '@odata.type': string },
): Entity {
  return createIntegrationEntity({
    entityData: {
      source: managedApp,
      assign: {
        _class: entities.MANAGED_APPLICATION._class,
        _type: entities.MANAGED_APPLICATION._type,
        // The key needs to be the name of the application so it can be looked up when making managed -> detected app relationships.
        // The managed app id is not available on the detected app so using it as the key here would make jobstate.findEntitiy not work.
        // The prefix is necessary to ensure key is at least 10 characters
        _key:
          MANAGED_APP_KEY_PREFIX + managedApp.displayName?.toLowerCase() ??
          managedApp.id, // Fallback to id if there is no name for the appds
        id: managedApp.id,
        name: managedApp.displayName?.toLowerCase(),
        displayName: managedApp.displayName as string,
        description: managedApp.description,
        notes: managedApp.notes ? [managedApp.notes] : [],
        COTS: !isLineOfBusiness(managedApp['@odata.type']),
        external: !isLineOfBusiness(managedApp['@odata.type']),
        mobile: isMobile(managedApp['@odata.type']),
        productionURL:
          (managedApp as WebApp).appUrl ??
          (managedApp as AndroidManagedStoreApp).appStoreUrl,
        publisher: managedApp.publisher,
        isPublished: managedApp.publishingState === 'published', // Essentially if it is available for download
        createdOn: parseTimePropertyValue(managedApp.createdDateTime),
        lastUpdatedOn: parseTimePropertyValue(managedApp.lastModifiedDateTime),
        featured: managedApp.isFeatured, // Indicates that they are featuring this app on their Company Portal
        privacyInformationURL: managedApp.privacyInformationUrl,
        informationURL: managedApp.informationUrl,
        owner: managedApp.owner || undefined, // Ex: Microsoft, Google, Facebook...
        developer: managedApp.developer, // Almost always the same as the owner

        // Line of Business Apps
        version: findNewestVersion(managedApp),
        committedContentVersion: (managedApp as MobileLobApp)
          .committedContentVersion,
        packageId: (managedApp as AndroidLobApp).packageId,
      },
    },
  });
}

/**
 * Creates an application entity that represents a global application.
 * This entity could be linked to applications in other integrations and therefore should not contain Intune-specific information.
 * https://docs.microsoft.com/en-us/graph/api/resources/intune-devices-detectedapp?view=graph-rest-beta
 */
export function createDetectedApplicationEntity(
  detectedApp: DetectedApp,
): Entity {
  return createIntegrationEntity({
    entityData: {
      source: detectedApp,
      assign: {
        _class: entities.DETECTED_APPLICATION._class,
        _type: entities.DETECTED_APPLICATION._type,
        // The key needs to be the name of the application so multiple relationships can be made to the same detected application entity.
        // The id is unique per detection so using it as the key would make jobstate.findEntitiy not work.
        // The prefix is necessary to ensure key is at least 10 characters
        _key:
          DETECTED_APP_KEY_PREFIX + detectedApp.displayName?.toLowerCase() ??
          detectedApp.id, // Fallback to id if there is no name for the app
        name: detectedApp.displayName?.toLowerCase(),
        displayName: detectedApp.displayName as string,
      },
    },
  });
}

/**
 * Line of business apps need to be manually uploaded to Azure ensuring that they are custom.
 * All other managed apps go throuhg an app store or a website (webApp).
 *
 * @param dataModel Microsoft datatype for the api response.
 * Examples: "#microsoft.graph.webApp", "#microsoft.graph.managedIOSStoreApp", "#microsoft.graph.androidLobApp"
 */
function isLineOfBusiness(dataModel: string) {
  return dataModel.toLowerCase().includes('lob');
}

/**
 *
 * @param dataModel Microsoft datatype for the api response.
 * Examples: "#microsoft.graph.androidStoreApp", "#microsoft.graph.managedIOSStoreApp", "#microsoft.graph.windowsMobileMSI", "#microsoft.graph.officeSuiteApp"
 */
function isMobile(dataModel: string) {
  return [
    'ios', // matches iPhone managed app types
    'android', // matches Android managed app types
    'mobile', // matches Windows phone app types
    'webApp', // webApps are availble for both mobile and desktop
  ].some((el) => dataModel.toLowerCase().indexOf(el) > -1);
}

export function findNewestVersion(
  managedApp: ManagedApp & WindowsPhoneXAP & AndroidLobApp & IosLobApp,
) {
  return (
    managedApp.versionName ?? // Version name is the most preferable version information as it is in the format X.X.X
    managedApp.versionCode ??
    managedApp.versionNumber ??
    managedApp.identityVersion ??
    managedApp.version ??
    UNVERSIONED
  );
}
