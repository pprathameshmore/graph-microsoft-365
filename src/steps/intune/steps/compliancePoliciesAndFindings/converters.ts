import {
  createIntegrationEntity,
  Entity,
  IntegrationLogger,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
import {
  DeviceCompliancePolicy,
  DeviceConfigurationDeviceStatus,
} from '@microsoft/microsoft-graph-types-beta';
import { entities } from '../../constants';
import { last } from 'lodash';
import {
  calculateSeverity,
  calculateNumericSeverity,
  findingIsOpen,
} from '../../utils';

/**
 * There are several different types of compliance policies based on your OS and what you are configuring.
 * So many that it does not make sense to ingest each individually until that granularity is desired
 * or necessary. Because of that, this method ingests all compliance policies the same way, and if
 * more granularity is desired, individual configuration settings can be viewed in the raw data.
 *
 * https://docs.microsoft.com/en-us/graph/api/resources/intune-shared-devicecompliancepolicy?view=graph-rest-beta
 */
export function createCompliancePolicyEntity(
  compliancePolicy: DeviceCompliancePolicy,
): Entity {
  return createIntegrationEntity({
    entityData: {
      source: compliancePolicy,
      assign: {
        _class: entities.COMPLIANCE_POLICY._class,
        _type: entities.COMPLIANCE_POLICY._type,
        function: 'endpoint-compliance', // needed to match up with existing integrations data structures.
        category: 'compliance', // needed for "ControlPolicy"
        id: compliancePolicy.id,
        name: compliancePolicy.displayName,
        description:
          compliancePolicy.description || compliancePolicy.displayName,
        createdOn: parseTimePropertyValue(compliancePolicy.createdDateTime),
        displayName: compliancePolicy.displayName,
        version: compliancePolicy.version,
        policyType: last(compliancePolicy['@odata.type']?.split('.')), // Possible values: 'iosCompliancePolicy'
      },
    },
  });
}

// https://docs.microsoft.com/en-us/graph/api/resources/intune-deviceconfig-deviceconfigurationdevicestatus?view=graph-rest-beta
export function createNoncomplianceFindingEntity(
  deviceStatus: DeviceConfigurationDeviceStatus,
  compliancePolicyEntity: Entity,
  logger: IntegrationLogger,
): Entity {
  return createIntegrationEntity({
    entityData: {
      source: deviceStatus,
      assign: {
        _class: entities.NONCOMPLIANCE_FINDING._class,
        _type: entities.NONCOMPLIANCE_FINDING._type,
        id: deviceStatus.id,
        name: 'Latest finding from ' + compliancePolicyEntity.displayName,
        category: 'endpoint',
        assessment: compliancePolicyEntity.displayName,
        status: deviceStatus.status, // Possible values are: unknown, notApplicable, compliant, remediated, nonCompliant, error, conflict, notAssigned.
        severity: calculateSeverity(deviceStatus.status, logger),
        numericSeverity: calculateNumericSeverity(deviceStatus.status, logger),
        open: findingIsOpen(deviceStatus.status, logger),
        lastProcessedOn: parseTimePropertyValue(
          deviceStatus.lastReportedDateTime,
        ),
        lastTestedOn: parseTimePropertyValue(deviceStatus.lastReportedDateTime),
        lastUpdatedOn: parseTimePropertyValue(
          deviceStatus.lastReportedDateTime,
        ),
      },
    },
  });
}
