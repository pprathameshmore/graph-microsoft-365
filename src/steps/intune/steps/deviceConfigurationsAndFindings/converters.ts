import {
  createIntegrationEntity,
  Entity,
  IntegrationLogger,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
import {
  DeviceConfiguration,
  DeviceConfigurationDeviceStatus,
} from '@microsoft/microsoft-graph-types-beta';
import { entities } from '../../constants';
import {
  calculateSeverity,
  calculateNumericSeverity,
  findingIsOpen,
} from '../../utils';
import { last } from 'lodash';

/**
 * There are several different types of device configurations based on your OS and what you are configuring.
 * So many that it does not make sense to ingest each individually until that granularity is desired
 * or necessary. Because of that, this method ingests all device configurations the same way, and if
 * more granularity is desired, individual configuration settings can be viewed in the raw data.
 *
 * https://docs.microsoft.com/en-us/graph/api/resources/intune-shared-deviceconfiguration?view=graph-rest-beta
 */
export function createDeviceConfigurationEntity(
  deviceConfiguration: DeviceConfiguration,
): Entity {
  return createIntegrationEntity({
    entityData: {
      source: deviceConfiguration,
      assign: {
        _class: entities.DEVICE_CONFIGURATION._class,
        _type: entities.DEVICE_CONFIGURATION._type,
        function: 'endpoint-configuration', // needed to match up with existing integrations data structures.
        category: 'config',
        id: deviceConfiguration.id,
        name: deviceConfiguration.displayName,
        description:
          deviceConfiguration.description || deviceConfiguration.displayName,
        createdOn: parseTimePropertyValue(deviceConfiguration.createdDateTime),
        displayName: deviceConfiguration.displayName,
        version: deviceConfiguration.version,
        /**
         * Possible values (There are many of them): https://docs.microsoft.com/en-us/graph/api/resources/intune-device-cfg-conceptual?view=graph-rest-beta
         * examples: iosCustomConfiguration, windows10GeneralConfiguration, iosWiFiConfiguration
         */
        configurationType: last(deviceConfiguration['@odata.type']?.split('.')),
      },
    },
  });
}

// https://docs.microsoft.com/en-us/graph/api/resources/intune-deviceconfig-deviceconfigurationdevicestatus?view=graph-rest-beta
export function createNoncomplianceFindingEntity(
  deviceStatus: DeviceConfigurationDeviceStatus,
  deviceConfigurationEntity: Entity,
  logger: IntegrationLogger,
): Entity {
  return createIntegrationEntity({
    entityData: {
      source: deviceStatus,
      assign: {
        _class: entities.NONCOMPLIANCE_FINDING._class,
        _type: entities.NONCOMPLIANCE_FINDING._type,
        id: deviceStatus.id,
        name: 'Latest finding from ' + deviceConfigurationEntity.displayName,
        category: 'endpoint',
        assessment: deviceConfigurationEntity.displayName,
        status: deviceStatus.status, // Possible values are: unknown, notApplicable, compliant, remediated, nonCompliant, error, conflict, notAssigned.
        severity: calculateSeverity(deviceStatus.status, logger),
        numericSeverity: calculateNumericSeverity(deviceStatus.status, logger),
        open: findingIsOpen(deviceStatus.status, logger),
        lastProcessedOn: parseTimePropertyValue(
          deviceStatus.lastReportedDateTime,
        ),
        lastTestedOn: parseTimePropertyValue(deviceStatus.lastReportedDateTime),
        lastUpdatedOn: parseTimePropertyValue(
          deviceStatus.lastReportedDateTime,
        ),
      },
    },
  });
}
