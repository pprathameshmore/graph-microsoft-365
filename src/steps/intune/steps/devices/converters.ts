import {
  createIntegrationEntity,
  createMappedRelationship,
  Entity,
  IntegrationError,
  Relationship,
  RelationshipDirection,
} from '@jupiterone/integration-sdk-core';
import {
  ComplianceState,
  DeviceType,
  ManagedDevice,
} from '@microsoft/microsoft-graph-types-beta';
import { entities as activeDirectoryEntities } from '../../../active-directory';
import { entities, INTUNE_HOST_AGENT_KEY_PREFIX } from '../../constants';
import { ManagedDeviceType, relationships } from '../../constants';

// https://docs.microsoft.com/en-us/graph/api/resources/intune-devices-manageddevice?view=graph-rest-1.0&viewFallbackFrom=graph-rest-beta
export function createManagedDeviceEntity(
  managedDevice: ManagedDevice,
): Entity {
  const isPhysicalDevice = isPhysical(managedDevice);
  const _class = ['Host'];
  if (isPhysicalDevice) {
    _class.push('Device');
  }
  return createIntegrationEntity({
    entityData: {
      source: managedDevice,
      assign: {
        _class,
        _type: selectDeviceType(managedDevice.deviceType, isPhysicalDevice),
        category: 'endpoint',
        id: managedDevice.id,
        name: managedDevice.deviceName,
        deviceName: managedDevice.deviceName,
        displayName: managedDevice.deviceName as string,
        deviceType: managedDevice.deviceType,
        model: managedDevice.model,
        make: managedDevice.manufacturer,
        version: managedDevice.model,
        serial: managedDevice.serialNumber,
        serialNumber: managedDevice.serialNumber,
        hardwareVendor:
          managedDevice.hardwareInformation?.manufacturer ??
          managedDevice.manufacturer,
        hardwareManufacturer:
          managedDevice.hardwareInformation?.manufacturer ??
          managedDevice.manufacturer,
        hardwareModel:
          managedDevice.hardwareInformation?.model ?? managedDevice.model,
        hardwareSerial:
          managedDevice.hardwareInformation?.serialNumber ??
          managedDevice.serialNumber,
        hardwareVersion:
          managedDevice.hardwareInformation?.model ?? managedDevice.model,
        meid: managedDevice.meid, // A mobile equipment identifier (MEID) is a globally unique number identifying a physical piece of CDMA2000 mobile station equipment
        imei: managedDevice.imei, // The International Mobile Equipment Identity (IMEI) is a number, usually unique, to identify 3GPP and iDEN mobile phones, as well as some satellite phones.
        iccid: managedDevice.iccid, // Every SIM card has a ICCID number, which stands for Integrated Circuit Card Identifier
        udid: managedDevice.udid, // UDID is an acronym for Unique Device Identifier.
        wifiMacAddress: managedDevice.wiFiMacAddress, // To communicate with a Wi-Fi network, a device must identify itself to the network using a unique network address called a Media Access Control (MAC) address
        easDeviceId: managedDevice.easActivated && managedDevice.easDeviceId, // Microsoft Exchange ActiveSync device identifier
        assetTag:
          managedDevice.azureActiveDirectoryDeviceId ??
          managedDevice.azureADDeviceId, // I believe these are legacy values now
        osDetails: `${managedDevice.operatingSystem} version ${managedDevice.osVersion}`,
        osName: managedDevice.operatingSystem,
        osVersion: managedDevice.osVersion,
        userEmails: managedDevice.emailAddress
          ? [managedDevice.emailAddress]
          : undefined,
        BYOD: managedDevice.managedDeviceOwnerType === 'personal',
        ownerType: managedDevice.managedDeviceOwnerType, // 'personal', 'company' or 'unknown'
        encrypted: managedDevice.isEncrypted,
        userId: managedDevice.userId,
        userDisplayName: managedDevice.userDisplayName,
        phoneNumber: managedDevice.phoneNumber,
        managed:
          managedDevice.managementState &&
          managedDevice.managementState !== 'discovered',
        supervised: managedDevice.isSupervised,
        jailBroken: managedDevice.jailBroken !== 'False',
        username: managedDevice.userPrincipalName,
        physical: isPhysicalDevice,
        // POTENTIAL: managedDevice.usersLoggedOn - link out to other users perhaps?
      },
    },
  });
}

export function createIntuneHostAgentEntity(
  managedDevice: ManagedDevice,
): Entity {
  return createIntegrationEntity({
    entityData: {
      source: managedDevice,
      assign: {
        _class: entities.HOST_AGENT._class,
        function: ['endpoint-compliance', 'endpoint-protection'], // https://github.com/JupiterOne/data-model/blob/master/src/schemas/HostAgent.json
        _type: entities.HOST_AGENT._type,
        _key: INTUNE_HOST_AGENT_KEY_PREFIX + managedDevice.id,
        name: 'intune-host-agent',
        displayName: 'Intune Host Agent',
        managementAgent: managedDevice.managementAgent, // Management channel of the device. Possible values include: eas, mdm, easMdm, intuneClient, easIntuneClient, jamf, googleCloudDevicePolicyController, ...
        state: managedDevice.managementState,
        registrationState: managedDevice.deviceRegistrationState,
        complianceState: managedDevice.complianceState,
        compliant: isCompliant(managedDevice.complianceState),
      },
    },
  });
}

/**
 * Converts Microsoft's compliance state to JupiterOne's `compliant` boolean.
 * There are some complianceStates where we can not determine if the device is compliant or not.
 * In those cases this key should remain `undefined`.
 */
function isCompliant(complianceState?: ComplianceState) {
  return ['unknown', 'configManager', undefined].includes(complianceState)
    ? undefined
    : complianceState === 'compliant';
}

/**
 * Is the device a physical device or a virtual server
 */
function isPhysical(managedDevice: ManagedDevice) {
  return managedDevice.model
    ? managedDevice.model !== 'Virtual Machine'
    : undefined; // https://docs.microsoft.com/en-us/mem/intune/fundamentals/windows-10-virtual-machines#reporting
}

/**
 * Select the correct device _type based on the Microsoft DeviceType
 * types determined from managed-questions-endpoint.yaml
 */
export function selectDeviceType(
  deviceType?: DeviceType,
  isPhysical?: Boolean,
): ManagedDeviceType {
  if (!isPhysical) {
    return 'server';
  }
  switch (deviceType) {
    case 'desktop':
    case 'mac':
    case 'unix':
    case 'linux':
      return 'desktop';
    case 'windowsRT': // retired mobile OS
    case 'winMO6': // retired mobile OS
    case 'nokia':
    case 'iPhone':
    case 'iPod':
    case 'android':
    case 'androidForWork': // same OS as android, idk why it is differentiated here
    case 'androidEnterprise': // same OS as android, idk why it is differentiated here
    case 'androidnGMS': // same OS as android, idk why it is differentiated here
    case 'blackberry':
    case 'palm':
      return 'smartphone';
    case 'winCE':
    case 'winEmbedded':
      return 'workstation';
    case 'macMDM':
    case 'windows10x':
      return 'laptop';
    case 'cloudPC':
      return 'server';
    case 'iSocConsumer': // some open source thing
    case 'holoLens': // augmented reality tool
    case 'surfaceHub': // smart whiteboard
    case 'unknown':
      return 'user_endpoint';
    default:
      return 'user_endpoint';
  }
}

/**
 * Creates a mapped relationship between a User and a Device if userId or email is provided.
 * Should be used when the active directory is not being ingested in this integration instance.
 */
export function createUserDeviceMappedRelationship(
  managedDeviceEntity: Entity,
  userId?: string,
  email?: string,
): Relationship | undefined {
  const relationship = relationships.MULTI_USER_HAS_DEVICE.find(
    (relationship) =>
      [relationship.sourceType, relationship.targetType].includes(
        managedDeviceEntity._type,
      ),
  );
  if (!relationship) {
    throw new IntegrationError({
      message: `Could not find relationship for deviceType ${managedDeviceEntity._type}`,
      code: 'RELATIONSHIP_IMPLEMENTATION_ERROR',
    });
  }
  return userId || email
    ? createMappedRelationship({
        _class: relationship._class,
        _type: relationship._type,
        _mapping: {
          sourceEntityKey: managedDeviceEntity._key,
          relationshipDirection: RelationshipDirection.REVERSE,
          targetFilterKeys: [
            ['_class', 'id'],
            ['_class', 'email'],
          ],
          targetEntity: {
            _class: activeDirectoryEntities.USER._class,
            _type: activeDirectoryEntities.USER._type,
            _key: userId,
            id: userId,
            email: email,
          },
        },
      })
    : undefined;
}
