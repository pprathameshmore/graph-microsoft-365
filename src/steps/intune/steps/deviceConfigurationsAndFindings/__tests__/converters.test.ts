import { createMockIntegrationLogger } from '@jupiterone/integration-sdk-testing';
import { createNoncomplianceFindingEntity } from '../converters';
import { exampleDeviceConfigurationEntity } from './fixtures/exampleDeviceConfigurationEntity';
import { exampleDeviceStatus } from './fixtures/exampleDeviceStatus';

const logger = createMockIntegrationLogger();

describe('createDeviceConfigurationEntity', () => {
  test('transfers properties correctly', () => {
    expect(
      createNoncomplianceFindingEntity(
        exampleDeviceStatus,
        exampleDeviceConfigurationEntity,
        logger,
      ),
    ).toMatchSnapshot('createNoncomplianceFindingEntity');
  });
});
