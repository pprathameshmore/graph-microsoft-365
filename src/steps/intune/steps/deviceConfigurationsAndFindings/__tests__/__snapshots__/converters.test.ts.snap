// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createDeviceConfigurationEntity transfers properties correctly: createNoncomplianceFindingEntity 1`] = `
Object {
  "_class": Array [
    "Finding",
  ],
  "_key": "7b521133-f056-40a9-b639-b76afd67e4bb_6aa605f0-744e-4ae1-8031-c6564624c89b_683dbff1-c8ff-4996-91ae-85484de46cfc",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "complianceGracePeriodExpirationDateTime": "2021-02-17T19:13:41.8113Z",
        "deviceDisplayName": "Personâ€™s iPhone",
        "deviceModel": null,
        "id": "7b521133-f056-40a9-b639-b76afd67e4bb_6aa605f0-744e-4ae1-8031-c6564624c89b_683dbff1-c8ff-4996-91ae-85484de46cfc",
        "lastReportedDateTime": "2021-03-01T12:20:19.1406851Z",
        "platform": 0,
        "status": "error",
        "userName": "dude.person@email.onmicrosoft.com",
        "userPrincipalName": "dude.person@email.onmicrosoft.com",
      },
    },
  ],
  "_type": "intune_noncompliance_finding",
  "active": undefined,
  "assessment": "Test iOS Wifi policy",
  "category": "endpoint",
  "createdOn": undefined,
  "displayName": "Latest finding from Test iOS Wifi policy",
  "id": "7b521133-f056-40a9-b639-b76afd67e4bb_6aa605f0-744e-4ae1-8031-c6564624c89b_683dbff1-c8ff-4996-91ae-85484de46cfc",
  "lastProcessedOn": 1614601219140,
  "lastTestedOn": 1614601219140,
  "lastUpdatedOn": 1614601219140,
  "name": "Latest finding from Test iOS Wifi policy",
  "numericSeverity": 6,
  "open": true,
  "severity": "high",
  "status": "error",
}
`;
